### PR TITLE
REGRESSION(279868@main) Alaska Airlines sign in popup loads behind the website's search field.

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### 73e553cb35754cc9fea44929e20199de4fa0cb6b
<pre>
REGRESSION(279868@main) Alaska Airlines sign in popup loads behind the website&apos;s search field.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279036">https://bugs.webkit.org/show_bug.cgi?id=279036</a>
&lt;<a href="https://rdar.apple.com/134911588">rdar://134911588</a>&gt;

Reviewed by Simon Fraser.

This webpage had multiple backing store providers, and content was incorrectly
added to the back one, despite overlapping the front one.

The overlap test uses the bounds of the provider, not the to-be-added layer to
check if it overlaps, and the providers themselves didn&apos;t overlap.

This restricts multiple backing store providers to only be used when they&apos;re
clipped (as was previously the case), so we can be sure the added layer doesn&apos;t
extend beyond the bounds of the provider. This shouldn&apos;t break the performance
improvement, since we still allow other composited layers to be added infront.

It does mean in some cases we keep the scroll clipped backing store provider
open, and prevent accumulating sharing layers into a further forward unclipped
backing provider. I think given the support for multiple open clipped providers,
this is a good tradeoff.

Ideally, we&apos;d allow accumulating bounds and adding to any provider, but that
seems like a riskier change, as we have to account for scrolling.

This also does a bit of cleanup, unifying the BackingSharingSnapshot and
preDescendantProviderStartLayer using a generation counter. It also adds a few
more comments, as I found the logic of why we end backing sharing sequences to
be hard to follow.

* LayoutTests/compositing/shared-backing/multiple-backing-sharing-providers-expected.html: Added.
* LayoutTests/compositing/shared-backing/multiple-backing-sharing-providers.html: Added.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::snapshot const):
(WebCore::RenderLayerCompositor::BackingSharingState::generation const):
(WebCore::RenderLayerCompositor::BackingSharingState::addBackingSharingCandidate):
(WebCore::RenderLayerCompositor::BackingSharingState::endBackingSharingSequence):
(WebCore::RenderLayerCompositor::BackingSharingState::backingProviderCandidateForLayer):
(WebCore::RenderLayerCompositor::computeCompositingRequirements):
(WebCore::RenderLayerCompositor::traverseUnchangedSubtree):
(WebCore::RenderLayerCompositor::updateBackingSharingBeforeDescendantTraversal):
(WebCore::RenderLayerCompositor::updateBackingSharingAfterDescendantTraversal):
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: <a href="https://commits.webkit.org/283515@main">https://commits.webkit.org/283515@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6e1f5d10a73f9e15192b1ebc2e976447318ec0fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66543 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19164 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70577 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17676 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53717 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17436 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/11924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69610 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57585 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33992 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38949 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16030 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72279 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/10500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/10532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57654 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60984 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14686 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/43985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42545 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderLayerCompositor.cpp [hunk 1]
```diff
@@ -284,12 +285,23 @@ struct RenderLayerCompositor::UpdateBackingTraversalState {
...
+
     To debug sharing behavior, enable the "Compositing" log channel and look for the P/p in the hierarchy output.
  */
 
+enum class BackingSharingSequenceIdentifierType { };
+using BackingSharingSequenceIdentifier = ObjectIdentifier<BackingSharingSequenceIdentifierType>;
+
+struct RenderLayerCompositor::BackingSharingSnapshot {
+    BackingSharingSequenceIdentifier sequenceIdentifier;
+    size_t providerCount;
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerCompositor.cpp]: Initialize to zero.

File: Source/WebCore/rendering/RenderLayerCompositor.cpp [hunk 2]
```diff
@@ -426,6 +446,13 @@ auto RenderLayerCompositor::BackingSharingState::backingProviderCandidateForLaye
 
     auto& candidate = m_backingProviderCandidates[candidateIndex];
 
+    // Only allow adding to providers that clip their descendants, unless there's only a single provider.
+    // Unclipped providers in-front are tracked for overlap testing only.
+    // FIXME: We could accumulate the union of the overlap bounds for a provider and its sharing layers to avoid this restriction.
+    if (m_backingProviderCandidates.size() > 1 && !candidate.providerLayer->canUseCompositedScrolling())
+        return nullptr;
+
+
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerCompositor.cpp]: Extra blank line.

File: Source/WebCore/rendering/RenderLayerCompositor.cpp [hunk 3]
```diff
@@ -337,6 +355,7 @@ class RenderLayerCompositor::BackingSharingState {
 
     Vector<Provider> m_backingProviderCandidates;
     RenderLayer* m_backingSharingStackingContext { nullptr };
+    BackingSharingSequenceIdentifier m_sequenceIdentifier { BackingSharingSequenceIdentifier::generate() };
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerCompositor.cpp:358]: It feels like an error to call `generate()` in two places.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/rendering/RenderLayerCompositor.cpp (hunk 3 above)
Comment (mattwoodrow) [reply Source/WebCore/rendering/RenderLayerCompositor.cpp:358]: In what sense? We initialise, and then update the identifier each time we end a sequence.

I guess I could make it only generate when we start a sequence, but then I'd need to make sure it gets set to invalid value instead at the end of a sequence (or make the getter return an optional I guess).
